### PR TITLE
Fix a use-after-free bug for detached threads

### PR DESCRIPTION
**Pull request by yamt**
the issue was pointed out by @alexcrichton in
https://github.com/WebAssembly/wasi-libc/issues/405

### CODE REVIEW

File: libc-top-half/musl/src/thread/pthread_create.c [hunk 1]
```diff
@@ -60,6 +60,18 @@ void __tl_sync(pthread_t td)
 	if (tl_lock_waiters) __wake(&__thread_list_lock, 1, 0);
 }
 
+#ifndef __wasilibc_unmodified_upstream
+/* this queue has at most one entry */
```
Comment (sbc100) [inline libc-top-half/musl/src/thread/pthread_create.c:64]: Give that, I'm not sure `queue` is the right name for this.

I'm not sure I have any better suggestions though.  Maybe `deferred_deallocation` and `process_deferred_deallocation`?

File: libc-top-half/musl/src/thread/pthread_create.c [hunk 2]
```diff
@@ -182,7 +194,17 @@ static void __pthread_exit(void *result)
...
-		free(self->map_base);
+		// As we use malloc/free which is considerably more complex
+		// than mmap/munmap to call and can even require a valid
+		// thread context, it's difficult to implement __unmapself.
+		//
+		// Here we take an alternative approach which simply defers
+		// the deallocation. An obvious downside of this approach is
+		// that it keeps the stack longer. (possibly forever.)
+		// To avoid wasting too much memory, we keep the free queue
+		// length at most one.
```
Comment (sbc100) [inline libc-top-half/musl/src/thread/pthread_create.c:205]: Can we be consistent about C vs C++ comments?  Below you've used C multi-line comment.

File: libc-top-half/musl/src/thread/pthread_create.c [hunk 3]
```diff
@@ -424,6 +446,16 @@ int __pthread_create(pthread_t *restrict res, const pthread_attr_t *restrict att
 			if (map == MAP_FAILED) goto fail;
 		}
 #else
+		/*
+		 * Process the free queue before allocationg a new one.
+		 * Hopefully it enables a reuse of the memory.
+		 *
+		 * Note: We can't perform a simple "handoff" becasue allocation
+		 * sizes might be different. (eg. the stack size might differ)
+		 */
```
Comment (sbc100) [inline libc-top-half/musl/src/thread/pthread_create.c:455]: The comment style within this file seems to be:

```
/* first line
  * second line
  * last line */
```

**Review by sbc100 — COMMENTED** [review]
lgtm % nits

File: libc-top-half/musl/src/thread/pthread_create.c (hunk 2 above)
Comment (yamt) [reply libc-top-half/musl/src/thread/pthread_create.c:205]: i followed nearby code.

File: libc-top-half/musl/src/thread/pthread_create.c (hunk 1 above)
Comment (yamt) [reply libc-top-half/musl/src/thread/pthread_create.c]: renamed

File: libc-top-half/musl/src/thread/pthread_create.c (hunk 2 above)
Comment (yamt) [reply libc-top-half/musl/src/thread/pthread_create.c]: fixed

File: libc-top-half/musl/src/thread/pthread_create.c (hunk 3 above)
Comment (yamt) [reply libc-top-half/musl/src/thread/pthread_create.c]: fixed

**Review by sbc100 — APPROVED** [review]

**Pull request merged**